File: app/src/main/java/org/evcs/android/features/main/MainActivity.java
package org.evcs.android.features.main;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;

import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.navigation.Navigation;
import androidx.viewpager.widget.ViewPager;

import com.base.core.adapter.viewpager.BaseFragmentStatePagerAdapter;
import com.google.android.material.bottomnavigation.BottomNavigationView;

import org.evcs.android.R;
import org.evcs.android.activity.AbstractSupportedVersionActivity;
import org.evcs.android.databinding.ActivityBaseNavhostWithBottomNavBinding;
import org.evcs.android.features.auth.initialScreen.AuthActivity;

import org.evcs.android.features.charging.ChargingTabFragment;
import org.evcs.android.features.map.MainMapFragment2;
import org.evcs.android.features.profile.ProfileFragment;
import org.evcs.android.features.shared.IVersionView;
import org.evcs.android.util.Extras;
import org.evcs.android.util.PushNotificationUtils;
import org.evcs.android.util.UserUtils;

public class MainActivity extends AbstractSupportedVersionActivity implements IVersionView {

    private BottomNavigationView mMenu;
    private ActivityResultLauncher mLoginResult;
    private ViewPager mViewPager;
    private BaseFragmentStatePagerAdapter mPagerAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mLoginResult = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(),
                (ActivityResultCallback) result -> { populate(); });
    }

    @Override
    @CallSuper
    protected void onPostCreate(@Nullable Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);

        createNotificationChannel();

        if (!getIntent().hasExtra(Extras.Root.VIEW_KEY)) {
            return;
        }
    }

    @Override
    protected View inflate(LayoutInflater layoutInflater) {
        @NonNull ActivityBaseNavhostWithBottomNavBinding binding =
                ActivityBaseNavhostWithBottomNavBinding.inflate(layoutInflater);
        setContentView(binding.getRoot());
        mMenu = binding.bottomNavigation;
        mViewPager = binding.activityBaseContent;
        return binding.getRoot();
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            String name = getString(R.string.notifications_channel_name);
            String description = getString(R.string.notifications_channel_description);
            int importance = NotificationManager.IMPORTANCE_DEFAULT;
            NotificationChannel channel = new NotificationChannel(PushNotificationUtils.CHANNEL_ID, name, importance);
            channel.setDescription(description);
            NotificationManager notificationManager = getSystemService(NotificationManager.class);
            notificationManager.createNotificationChannel(channel);
        }
    }

    @Override
    public void isSupportedVersion(boolean isSupported, String versionWording) {
        if(!isSupported) {
            showNotSupportedVersion(versionWording);
        }
    }

    public BottomNavigationView getMenuView() {
        return mMenu;
    }

    @Override
    protected void populate() {
        super.populate();
        mMenu.getMenu().findItem(R.id.menu_drawer_profile).setTitle((UserUtils.getLoggedUser() == null) ?
                "Sign in" :
                getString(R.string.drawer_menu_profile));
        mMenu.setSelectedItemId(R.id.menu_drawer_map);
        mPagerAdapter = new BaseFragmentStatePagerAdapter(getSupportFragmentManager());
        MainMapFragment2 mapFragment = MainMapFragment2.Companion.newInstance();
        mPagerAdapter.addItem(mapFragment);
        ChargingTabFragment chargingTabFragment = ChargingTabFragment.Companion.newInstance();
        mPagerAdapter.addItem(chargingTabFragment);
        ProfileFragment profileFragment = ProfileFragment.Companion.newInstance();
        mPagerAdapter.addItem(profileFragment);
        mViewPager.setAdapter(mPagerAdapter);
    }

    @Override
    protected void setListeners() {
        super.setListeners();
        mMenu.getMenu().findItem(R.id.menu_drawer_profile).setTitle((UserUtils.getLoggedUser() == null) ?
                "Sign in" :
                getString(R.string.drawer_menu_profile));
        mMenu.setOnNavigationItemSelectedListener(new BottomNavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.menu_drawer_map:
                        mViewPager.setCurrentItem(0);
                        break;
                    case R.id.menu_drawer_charging:
                        if (UserUtils.getLoggedUser() == null)
                            mLoginResult.launch(new Intent(MainActivity.this, AuthActivity.class));
                        else
                            mViewPager.setCurrentItem(1);
                        break;
                    case R.id.menu_drawer_profile:
                        if (UserUtils.getLoggedUser() == null)
                            mLoginResult.launch(new Intent(MainActivity.this, AuthActivity.class));
                        else
                            mViewPager.setCurrentItem(2);

                }
                return true;
            }
        });
    }

    @Override
    public void onNetworkError() {
        //Handled by fragment
    }

}

File: app/src/main/java/org/evcs/android/features/main/MainActivity.kt
package org.evcs.android.features.main

import android.app.NotificationChannel
import android.app.NotificationManager
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import androidx.annotation.CallSuper
import androidx.core.view.isVisible
import androidx.navigation.Navigation.findNavController
import com.base.core.util.NavigationUtils
import com.google.android.material.bottomnavigation.BottomNavigationView
import org.evcs.android.R
import org.evcs.android.activity.AbstractSupportedVersionActivity
import org.evcs.android.databinding.ActivityBaseNavhostWithBottomNavBinding
import org.evcs.android.features.profile.plans.PlansActivity
import org.evcs.android.features.shared.IVersionView
import org.evcs.android.util.Extras
import org.evcs.android.util.PushNotificationUtils
import org.evcs.android.util.UserUtils

class MainActivity : AbstractSupportedVersionActivity(), IVersionView {
    var mNavigationController: MainNavigationController? = null
    private lateinit var menuView: BottomNavigationView
    private lateinit var mButton: TextView
    var isBottomOfStack = true
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isBottomOfStack = intent.getBooleanExtra(Extras.MainActivity.IS_BOTTOM, true)
    }

    @CallSuper
    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        createNotificationChannel()
        mNavigationController = MainNavigationController(this, true,
            findNavController(this, R.id.activity_base_content))
        //mNavigationController!!.startFlow()

//        if (!intent.hasExtra(Extras.Root.VIEW_KEY)) {
//            return
//        }
    }

    override fun inflate(layoutInflater: LayoutInflater): View {
        val binding = ActivityBaseNavhostWithBottomNavBinding.inflate(layoutInflater)
        setContentView(binding.root)
        menuView = binding.bottomNavigation
        mButton = binding.bottomNavigationButton
        return binding.root
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name = getString(R.string.notifications_channel_name)
            val description = getString(R.string.notifications_channel_description)
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val channel = NotificationChannel(PushNotificationUtils.CHANNEL_ID, name, importance)
            channel.description = description
            val notificationManager = getSystemService(NotificationManager::class.java)
            notificationManager.createNotificationChannel(channel)
        }
    }

    override fun getNavGraphId(): Int {
        return R.navigation.navigation_graph
    }

    override fun isSupportedVersion(isSupported: Boolean, versionWording: String?) {
        if (!isSupported) {
            showNotSupportedVersion(versionWording)
        }
    }

    override fun populate() {
        super.populate()
        if (UserUtils.getLoggedUser() == null) {
            menuView.menu.clear()
            mButton.visibility = View.VISIBLE
        } else {
            menuView.inflateMenu(R.menu.drawer)
            mButton.visibility = View.GONE
        }
        menuView.selectedItemId = R.id.menu_drawer_map
        updateProfileAlert()
    }

    override fun setListeners() {
        super.setListeners()
        mButton.setOnClickListener { NavigationUtils.jumpTo(this, PlansActivity::class.java) }
        menuView.setOnNavigationItemSelectedListener { item ->
            when (item.itemId) {
                R.id.menu_drawer_map -> mNavigationController!!.onMapClicked()
                R.id.menu_drawer_charging -> mNavigationController!!.goToPreCharging()
                R.id.menu_drawer_profile -> mNavigationController!!.goToProfile()
            }
            //will be updated after canceling session
            false
        }
    }

    override fun onNetworkError() {
        //Handled by fragment
    }

    fun setSelectedItem(item: Int) {
        for (i in 0 until menuView.menu.size()) {
            menuView.menu.getItem(i).isChecked = false
        }
        menuView.menu.findItem(item)?.isChecked = true
    }

    fun updateProfileAlert() {
//        showProfileAlert(UserUtils.getLoggedUser()?.activeSubscription?.issue ?: false ||
//        !(UserUtils.getLoggedUser()?.isPhoneVerified ?: true))
    }

    fun showProfileAlert(show: Boolean) {
        if (!::menuView.isInitialized) return
        if (show) {
            menuView.getOrCreateBadge(R.id.menu_drawer_profile)
        } else {
            menuView.removeBadge(R.id.menu_drawer_profile)
        }
    }

    //Needed because if there's a scrollview the menu can be shown above the keyboard instead of below
    fun attachKeyboardListener() {
        val view = window.decorView.rootView
        KeyboardListener.attach(view) { isShown -> menuView.isVisible = !isShown }
    }

    fun detachKeyboardListener() {
        menuView.isVisible = true
        KeyboardListener.detach(window.decorView.rootView)
    }

}